### PR TITLE
Extend Jepsen tests with support for server

### PR DESCRIPTION
**Pull request by antonio2368**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add support for testing ClickHouse server with Jepsen. By the way, we already have support for testing ClickHouse Keeper with Jepsen. This pull request extends it to Replicated tables.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: tests/jepsen.clickhouse-keeper/src/jepsen/clickhouse/server/client.clj [hunk 1]
```diff
@@ -0,0 +1,49 @@
+(ns jepsen.clickhouse.server.client
+  (:require [clojure.java.jdbc :as j]
+            [jepsen.reconnect :as rc]))
+
+(defn db-spec
```
Comment (antonio2368) [inline tests/jepsen.clickhouse-keeper/src/jepsen/clickhouse/server/client.clj:5]: biggest unknown was which client to use, but jdbc works well 😄

**Review by alesapin — APPROVED** [review]
Great job! The only extension I'd like to have is dynamic cluster configuration or nemesis which adds and remove nodes constantly (both for keeper and server).

File: tests/jepsen.clickhouse-keeper/src/jepsen/clickhouse/server/set.clj [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+          (j/query c "CREATE TABLE set ON CLUSTER test_cluster (value Int64) Engine=ReplicatedMergeTree ORDER BY value")))))
+
+  (invoke! [this test op]
+    (chc/with-exception op
+      (chc/with-connection [c conn] (= :read (:f op))
+        (case (:f op)
+          :add (do
+                  (j/query c (str "INSERT INTO set VALUES (" (:value op) ")"))
+                  (assoc op :type :ok))
+          :read (->> (j/query c "SELECT value FROM set")
```
[the remark this answers is not in the document]
Comment (antonio2368) [reply tests/jepsen.clickhouse-keeper/src/jepsen/clickhouse/server/set.clj]: It just checks for extra data or data loss, so no order required here

File: tests/jepsen.clickhouse/src/jepsen/clickhouse/main.clj [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+                   ; interesting here?
+                   :exit  (.getExitStatus cmd))))
+        (finally
+          (.release semaphore)))))
+
+  (upload! [this ctx local-paths remote-path _opts]
+    (when (:dummy conn-spec)
+      (throw+ {:type :jepsen.control/dummy}))
+    (with-errors conn-spec ctx
+      (with-open [sftp (.newSFTPClient client)]
```
[the remark this answers is not in the document]
Comment (antonio2368) [reply tests/jepsen.clickhouse/src/jepsen/clickhouse/main.clj:78]: https://github.com/jepsen-io/jepsen/pull/554
Just this until it's released officially.

File: tests/jepsen.clickhouse/src/jepsen/clickhouse/server/set.clj [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        (chc/with-connection [c conn] false
+          (j/query c "DROP TABLE IF EXISTS set ON CLUSTER test_cluster")
+          (j/query c "CREATE TABLE set ON CLUSTER test_cluster (value Int64) Engine=ReplicatedMergeTree ORDER BY value")))))
+
+  (invoke! [this test op]
+    (chc/with-exception op
+      (chc/with-connection [c conn] (= :read (:f op))
+        (case (:f op)
+          :add (do
+                  (j/query c (str "INSERT INTO set VALUES (" (:value op) ")"))
```
[the remark this answers is not in the document]
Comment (antonio2368) [reply tests/jepsen.clickhouse/src/jepsen/clickhouse/server/set.clj:31]: and inserts need ` insert_quorum=auto, insert_quorum_parallel=0`?
It seems it passes most of the time locally but sometimes there are stale elements.

**Pull request merged**